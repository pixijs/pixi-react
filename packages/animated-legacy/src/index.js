export * from '@pixi/react-pixi-animated/src/index';
