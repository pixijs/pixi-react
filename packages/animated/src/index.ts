import { TYPES } from '@pixi/react';

import { animated } from './react-spring';

export const BitmapText = animated[TYPES.BitmapText];
export const Container = animated[TYPES.Container];
export const Graphics = animated[TYPES.Graphics];
export const NineSlicePlane = animated[TYPES.NineSlicePlane];
export const ParticleContainer = animated[TYPES.ParticleContainer];
export const Sprite = animated[TYPES.Sprite];
export const AnimatedSprite = animated[TYPES.AnimatedSprite];
export const Text = animated[TYPES.Text];
export const TilingSprite = animated[TYPES.TilingSprite];
export const SimpleMesh = animated[TYPES.SimpleMesh];
export const SimpleRope = animated[TYPES.SimpleRope];
