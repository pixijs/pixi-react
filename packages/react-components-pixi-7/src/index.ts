/// <reference path="./global.d.ts" />
import type { ComponentsType, PixiComponentType } from '@pixi/react-types';

import {
    AnimatedSprite as AnimatedSpriteComponent,
    BitmapText as BitmapTextComponent,
    Container as ContainerComponent,
    Graphics as GraphicsComponent,
    NineSlicePlane as NineSlicePlaneComponent,
    ParticleContainer as ParticleContainerComponent,
    SimpleMesh as SimpleMeshComponent,
    SimpleRope as SimpleRopeComponent,
    Sprite as SpriteComponent,
    Text as TextComponent,
    TilingSprite as TilingSpriteComponent,
} from './components';
import { applyDefaultProps } from './utils/props';

import type {
    PixiReactContainer,
    ReactStageComponent,
    ReactAnimatedSpriteComponent,
    ReactBitmapTextComponent,
    ReactContainerComponent,
    ReactGraphicsComponent,
    ReactNineSlicePlaneComponent,
    ReactParticleContainerComponent,
    ReactSimpleMeshComponent,
    ReactSimpleRopeComponent,
    ReactSpriteComponent,
    ReactTextComponent,
    ReactTilingSpriteComponent,
} from './types';

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export { Context as AppContext, AppProvider, AppConsumer, withPixiApp } from './stage/provider';
export { useTick, useApp } from './hooks';
export { withFilters } from './hoc';
export { eventHandlers } from './utils/pixi';
export { configurePixiReactStage } from './stage';
export { configurePixiReactRenderAPI } from './render';
export { applyDefaultProps };
export type { PixiReactContainer, ReactStageComponent };

/**
 * Available tag types
 *
 * @type {Object}
 */
export const TYPES = {
    BitmapText: 'BitmapText',
    Container: 'Container',
    Graphics: 'Graphics',
    NineSlicePlane: 'NineSlicePlane',
    ParticleContainer: 'ParticleContainer',
    Sprite: 'Sprite',
    AnimatedSprite: 'AnimatedSprite',
    Text: 'Text',
    TilingSprite: 'TilingSprite',
    SimpleMesh: 'SimpleMesh',
    SimpleRope: 'SimpleRope',
};

export const BitmapText = TYPES.BitmapText as unknown as ReactBitmapTextComponent;
export const Container = TYPES.Container as unknown as ReactContainerComponent;
export const Graphics = TYPES.Graphics as unknown as ReactGraphicsComponent;
export const NineSlicePlane = TYPES.NineSlicePlane as unknown as ReactNineSlicePlaneComponent;
export const ParticleContainer = TYPES.ParticleContainer as unknown as ReactParticleContainerComponent;
export const Sprite = TYPES.Sprite as unknown as ReactSpriteComponent;
export const AnimatedSprite = TYPES.AnimatedSprite as unknown as ReactAnimatedSpriteComponent;
export const Text = TYPES.Text as unknown as ReactTextComponent;
export const TilingSprite = TYPES.TilingSprite as unknown as ReactTilingSpriteComponent;
export const SimpleMesh = TYPES.SimpleMesh as unknown as ReactSimpleMeshComponent;
export const SimpleRope = TYPES.SimpleRope as unknown as ReactSimpleRopeComponent;

const components: ComponentsType = {
    AnimatedSprite: AnimatedSpriteComponent,
    BitmapText: BitmapTextComponent,
    Container: ContainerComponent,
    Graphics: GraphicsComponent,
    NineSlicePlane: NineSlicePlaneComponent,
    ParticleContainer: ParticleContainerComponent,
    SimpleMesh: SimpleMeshComponent,
    SimpleRope: SimpleRopeComponent,
    Sprite: SpriteComponent,
    Text: TextComponent,
    TilingSprite: TilingSpriteComponent,
};

export function configurePixiReactComponents(
    PixiComponent: PixiComponentType,
)
{
    // Register default components in `react-modular`'s COMPONENTS hash
    Object.keys(TYPES).forEach((type) => PixiComponent(type, components[type]));

    return {
        TYPES,
        applyDefaultProps,
    };
}
