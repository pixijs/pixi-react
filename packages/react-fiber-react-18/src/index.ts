import Reconciler from 'react-reconciler';
import type { applyPropsType, ComponentType, PropsType, PixiReactMinimalExpandoContainer } from '@pixi/react-types';
import data from './data.json';
import { makeHostConfig } from './hostconfig';
import type { diffPropertiesType, PixiReactHostConfig, PixiReactReconciler } from './types';

export type { PixiReactHostConfig, PixiReactReconciler };

export function configurePixiReactHostConfig<
    Container extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer,
>({
    COMPONENTS,
    applyDefaultProps,
    diffProperties,
}: {
    COMPONENTS: Record<string, ComponentType<PropsType, Container, Instance>>;
    applyDefaultProps: applyPropsType<PropsType, Instance>;
    diffProperties?: diffPropertiesType<Instance>;
})
{
    return makeHostConfig<Container, Instance>({ COMPONENTS, applyDefaultProps, diffProperties });
}

export function configurePixiReactFiber<
    PixiContainer extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer,
>(hostConfig: PixiReactHostConfig<PixiContainer, Instance>): PixiReactReconciler<PixiContainer, Instance>
{
    const PixiReactFiber = Reconciler(hostConfig);

    PixiReactFiber.injectIntoDevTools({
        bundleType: process.env.NODE_ENV !== 'production' ? 1 : 0,
        version: data.REACT_DOM_VERSION,
        rendererPackageName: data.PACKAGE_NAME,
    });

    return PixiReactFiber;
}
