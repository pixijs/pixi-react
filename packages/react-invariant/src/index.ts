// borrowed from fbjs
export function invariant(condition: boolean, format: string, ...args: string[])
{
    if (process.env.NODE_ENV === 'production')
    {
        return;
    }

    if (!condition)
    {
        let error;

        if (format === undefined)
        {
            error = new Error(
                'Minified exception occurred; use the non-minified dev environment '
                + 'for the full error message and additional helpful warnings.'
            );
        }
        else
        {
            let argIndex = 0;

            error = new Error(format.replace(/%s/g, () => String(args[argIndex++])));
            error.name = 'Invariant Violation';
        }

        // @ts-ignore: as per original fbjs implementation
        error.framesToPop = 1; // Skip invariant's own stack frame.
        throw error;
    }
}
