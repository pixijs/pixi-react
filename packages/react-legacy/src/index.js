export * from '@pixi/react/src/index';
