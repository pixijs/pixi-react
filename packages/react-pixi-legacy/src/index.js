export * from '@pixi/react-pixi/src/index';
