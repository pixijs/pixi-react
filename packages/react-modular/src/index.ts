import { invariant } from '@pixi/react-invariant';
import type {
    applyPropsType,
    ComponentType,
    MinimalPixiReactFiber,
    PixiComponentType,
    PixiReactMinimalExpandoContainer,
    PropsType,
    Roots,
    CreateRootType,
    RenderType,
    UnmountComponentAtNodeType,
    MinimalContainer,
    MinimalHostConfig,
    ComponentsType,
} from '@pixi/react-types';

type ConfigurePixiReactHostConfigType<
    Container extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer,
    HostConfigType extends MinimalHostConfig<Container, Instance>,
> = (params: {
    COMPONENTS: Record<string, ComponentType<PropsType, Container, Instance>>;
    applyDefaultProps: applyPropsType<PropsType, Instance>;
}) => HostConfigType;

type ConfigurePixiReactFiberType<
    PixiContainer extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer,
    HostConfigType extends MinimalHostConfig<PixiContainer, Instance>,
    PixiReactFiberType extends MinimalPixiReactFiber<PixiContainer>,
> = (hostConfig: HostConfigType) => PixiReactFiberType;

type ConfigurePixiReactComponentsType<Instance extends PixiReactMinimalExpandoContainer> = (
    pixiComponent: PixiComponentType,
) => {
    TYPES: Record<string, string>;
    applyDefaultProps: applyPropsType<PropsType, Instance>;
};

type ConfigureStageType<
    ConcreteStageType,
    Container extends MinimalContainer,
    PixiReactFiberType extends MinimalPixiReactFiber<Container>,
> = (pixiReactFiber: PixiReactFiberType) => ConcreteStageType;

type ConfigurePixiReactRenderAPIType<
    Container extends MinimalContainer,
    PixiReactFiberType extends MinimalPixiReactFiber<Container>,
> = (pixiReactFiber: PixiReactFiberType) => {
    roots: Roots<Container>;
    createRoot: CreateRootType<Container>;
    render: RenderType<Container>;
    unmountComponentAtNode: UnmountComponentAtNodeType<Container>;
};

export function configurePixiComponent()
{
    const COMPONENTS: ComponentsType = {};

    /**
     * Create Component
     *
     * @param {string} type
     * @param {Object} lifecycle methods
     */
    function PixiComponent<
        P extends PropsType,
        Container extends PixiReactMinimalExpandoContainer,
        Instance extends PixiReactMinimalExpandoContainer,
    >(type: string, lifecycle: ComponentType<P, Container, Instance>)
    {
        invariant(!!type, 'Expect type to be defined, got `%s`', type);
        invariant(
            !COMPONENTS[type],
            'Component `%s` could not be created, a component with that name already exists.',
            type,
        );

        COMPONENTS[type] = lifecycle;

        return type as unknown as React.ComponentType<P>;
    }

    return { COMPONENTS, PixiComponent };
}

export function configurePixiReact<
    ConcreteStageType,
    Container extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer,
    HostConfigType extends MinimalHostConfig<Container, Instance>,
    PixiReactFiberType extends MinimalPixiReactFiber<Container>,
>({
    configurePixiReactHostConfig,
    configurePixiReactFiber,
    configurePixiReactComponents,
    configurePixiReactStage,
    configurePixiReactRenderAPI,
}: {
    configurePixiReactHostConfig: ConfigurePixiReactHostConfigType<Container, Instance, HostConfigType>;
    configurePixiReactFiber: ConfigurePixiReactFiberType<Container, Instance, HostConfigType, PixiReactFiberType>;
    configurePixiReactComponents: ConfigurePixiReactComponentsType<Instance>;
    configurePixiReactStage: ConfigureStageType<ConcreteStageType, Container, PixiReactFiberType>;
    configurePixiReactRenderAPI: ConfigurePixiReactRenderAPIType<Container, PixiReactFiberType>;
})
{
    const { COMPONENTS, PixiComponent } = configurePixiComponent();

    // `configurePixiReactComponents` depends on this module's `PixiComponent` in order to build and store references to
    // instances of the default PixiReact component types
    const {
        // TODO: do we actually need to return this here, could just be imported by consumer, it depends if packages might
        // want to create more than the current set of TYPES
        TYPES,
        // TODO: same here... also consider moving applyDefaultProps to its own package
        applyDefaultProps,
    } = configurePixiReactComponents(PixiComponent);

    // Most of the React Reconciler is static, the exceptions are:
    // 1) `createElement` depends on `COMPONENTS` which is defined above in this module `react-modular` but configured by
    //    the previous step `configurePixiReactComponents`
    // 2) `commitUpdate` depends on `applyDefaultProps` which comes from the previous `configurePixiReactComponents` step
    const hostConfig = configurePixiReactHostConfig({
        COMPONENTS,
        applyDefaultProps,
    });
    const PixiReactFiber = configurePixiReactFiber(hostConfig);

    const Stage = configurePixiReactStage(PixiReactFiber);

    // `configurePixiReactRenderAPI` needs access to a `PixiReactFiber` instance in order to build the
    // `createRoot` / `render` API
    const { createRoot, render, unmountComponentAtNode } = configurePixiReactRenderAPI(PixiReactFiber);

    return {
        // From this module
        COMPONENTS,
        PixiComponent,
        // From pixi-react-fiber
        PixiReactFiber,
        // From pixi-react-components
        TYPES,
        createRoot,
        render,
        unmountComponentAtNode,
        Stage,
    };
}
