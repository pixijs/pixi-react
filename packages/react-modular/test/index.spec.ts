import * as index from '../src/index';
const { configurePixiComponent } = index;

describe('index', () =>
{
    test('export modules for react-modular', () =>
    {
        expect(index).toMatchSnapshot();
    });
});

describe('PixiComponent', () =>
{
    test('type must be defined', () =>
    {
        const { PixiComponent } = configurePixiComponent();

        // @ts-ignore: specifically testing untyped case
        expect(() => PixiComponent(null, { create: () => {} })).toThrow('Expect type to be defined, got `null`');
    });

    test('cannot override existing component', () =>
    {
        const { PixiComponent } = configurePixiComponent();

        // @ts-ignore - ignore for test case
        PixiComponent('Text', { create: () => {} });

        // @ts-ignore - ignore for test case
        expect(() => PixiComponent('Text', { create: () => {} })).toThrow(
            'Component `Text` could not be created, a component with that name already exists.',
        );
    });

    test('inject custom component', () =>
    {
        const { COMPONENTS, PixiComponent } = configurePixiComponent();

        const lifecycle = { create: () => {} };

        // @ts-ignore - ignore for test case
        PixiComponent('Rectangle', lifecycle);
        expect(COMPONENTS).toHaveProperty('Rectangle', lifecycle);
    });
});
