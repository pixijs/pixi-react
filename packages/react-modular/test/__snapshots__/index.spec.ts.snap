// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index export modules for react-modular 1`] = `
{
  "configurePixiComponent": [Function],
  "configurePixiReact": [Function],
}
`;
